sqlglot/optimizer/qualify_columns.py

```diff
@@ -380,18 +410,109 @@ def _qualify_columns(scope: Scope, resolver: Resolver) -> None:
+    if not dot_column.is_type(exp.DataType.Type.STRUCT):
+        return []
+
+    # All nested struct values are ColumnDefs, so normalize the first exp.Column in one
+    dot_column = dot_column.copy()
+    starting_struct = exp.ColumnDef(this=dot_column.this, kind=dot_column.type)
+
+    # First part is the table name and last part is the star so they can be dropped
+    dot_parts = expression.parts[1:-1]
+
+    # If we're expanding a nested struct (eg. t.c.f1.f2.*), start the DFS traversal from that struct
+    for part in dot_parts[1:]:
+        struct_fields = t.cast(exp.DataType, starting_struct.kind).expressions
+
+        if not struct_fields:
+            return []
+
+        for fld in struct_fields:
+            # Unable to expand star unless all struct fields are named
+            if not isinstance(fld.this, exp.Identifier):
+                return []
+
+            if fld.name == part.name and fld.kind.this == exp.DataType.Type.STRUCT:
+                starting_struct = fld
+
+    if starting_struct.name != dot_parts[-1].name:
+        return []
+
+    # Each nested field caches its path in exp.Identifier parts e.g. tbl.f0.f1.item -> [exp.Identifier(tbl), ..., exp.Identifier(item)]
+    nested_fields: t.Dict[int, t.List[exp.Identifier]] = {id(starting_struct): []}
+
+    # The keys to non-struct field entries in `nested_fields`
+    nested_field_keys = []
+
+    stack = [(starting_struct, id(starting_struct))]
+
+    # Traverse & flatten the starting struct
+    while stack:
+        fld, parent_struct_id = stack.pop()
+
+        # Unable to expand star unless all struct fields are named and annotated
+        if not isinstance(fld.this, exp.Identifier) or not fld.kind:
+            return []
+
+        id_fld = id(fld)
+
+        # Preserve the qualified path for this field
+        nested_fields[id_fld] = list(
+            itertools.chain(
+                [part.copy() for part in nested_fields[parent_struct_id]], [fld.this.copy()]
+            )
+        )
+
+        if not fld.kind.is_type(exp.DataType.Type.STRUCT):
+            # Collect the non-struct columns
+            nested_field_keys.append(id_fld)
+        else:
+            stack.extend((expr, id_fld) for expr in reversed(fld.kind.expressions))
+
+    new_selections = []
+    for nested_fields_key in nested_field_keys:
+        # The fully qualified path of each field is created by:
+        # - The path until the starting struct, if any (dot_parts)
+        # - The path from the starting struct until the nested field (nested_structs[key])
+        root, *parts = list(
+            itertools.chain(
+                [part.copy() for part in dot_parts[:-1]], nested_fields[nested_fields_key]
+            )
+        )
+        new_column: exp.Column | exp.Dot = exp.column(
+            t.cast(exp.Identifier, root), table=dot_column.table
+        )
+
+        if parts:
+            new_column = exp.Dot.build([new_column, *parts])
+
```

Comment:
Maybe this can be simplified to `new_column = exp.column(root, table=..., fields=parts)`.